### ISSUE TITLE
Confusing undefined function error for Code.eval_quoted_with_env

### ISSUE THREAD

**Issue by sabiwara** [elixir-lang/elixir#12982]
### Elixir and Erlang/OTP versions

```
Erlang/OTP 26 [erts-14.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]

Elixir 1.15.6 (compiled with Erlang/OTP 26)
```

Also on `main`

### Operating system

MacOS

### Current behavior

```elixir
iex> defmodule X, do: (def env, do: __ENV__)
iex> Code.eval_quoted_with_env(quote(do: y()), [], X.env())
** (UndefinedFunctionError) function :erl_eval.y/0 is undefined or private
    (stdlib 5.1) :erl_eval.y/0
```

### Expected behavior

Have something consistent with

```elixir
iex> Code.eval_quoted_with_env(quote(do: y()), [], __ENV__)
error: undefined function y/0 (there is no such import)
```

or at least not refer to the `:erl_eval` module.

Will look into this later, registering so I don't forget.

**Comment by josevalim** [comment]
@sabiwara if it helps, the erl_eval module has a concept of local_handler, which we could override/pass as argument to emit a custom error :)

**Comment by sabiwara** [comment]
@josevalim thank you, it seems exactly what I needed!

I started looking into this earlier today, here are my findings so far:
- when using a regular `Code.eval_quoted` with `__ENV__`, `function == nil` so we generate the undefined error [during expansion](https://github.com/elixir-lang/elixir/blob/main/lib/elixir/src/elixir_expand.erl#L851-L852)
- when `env.function != nil`, we emit the following erlang for a local call: `{:call, 1, {:atom, 1, :y}, []}`
- `:erl_eval.expr({:call, 1, {:atom, 1, :y}, []},%{},:none)` raises `** (UndefinedFunctionError) function :erl_eval.y/0 is undefined or private`
- for regular compiled module code (the usual case where `function != nil`), [`:elixir_module`](https://github.com/elixir-lang/elixir/blob/58bb5194d3cb1c4e702b0b1b0d065b1f551b9651/lib/elixir/src/elixir_module.erl#L144) would cause a compilation error and the above will never get executed
- the particular (unnatural?) combination of `eval_quoted` and a tweaked env with `function != nil` causes this

I'm not sure what the best fix would be, but I was considering the following two things:
- fix `:elixir.eval_forms` - as you mentioned, add a local_handler to `erl_eval` seems be perfect for this! PR upcoming
- update [`Macro.Env`](https://hexdocs.pm/elixir/Macro.Env.html) doc recipe to return `%{__ENV__ | function: nil, module: nil}` or add a `Macro.Env.new/0` constructor wrapping this. I think it might be a good thing disregarding the fix above to avoid the unnatural configuration of evaluating within a function without a module compilation.

**Comment by josevalim** [comment]
I see. This is a very specific case of passing a certain `__ENV__` to `Code.eval_quoted_with_env`, otherwise we are generally ok. I assume this happens during pry? This is tricky, because we would expand prying a function and then doing `__ENV__` returns both module and function, so we can't precisely prune it.

**Comment by josevalim** [comment]
TL;DR: I agree with local handler being the way to go.

**Comment by sabiwara** [comment]
> I assume this happens during pry?

I don't know, my use case was simply trying to follow this [suggestion to "make a custom env"](https://hexdocs.pm/elixir/main/Macro.Env.html). Which I think might be sub-optimal since it creates it within a module's function and therefore not so realistic to evaluate code.

> TL;DR: I agree with local handler being the way to go.

I think I got it! https://github.com/elixir-lang/elixir/pull/12998
Thanks again for the hint, you really unblocked me ❤️
